NtupleProducer/plugins/bRegressionProducer.cc

```diff
@@ -0,0 +1,570 @@
+                chEnergies[icone] += candEnergy;
+            } else {
+                neEnergies[icone] += candEnergy;
+            }
+        }
+        
+        if (abs(softLepPDGId)==13){
+            isMu=1; 
+        }else if (abs(softLepPDGId)==11){
+            isEle=1;
+        }else{
+            isOther=1;
+        }
+
+
+        float maxFoundSignificance = 0;
+        VertexDistance3D vdist;
+        int NSV = 0;
+        const auto& pv = vtxProd.at(0);
+        for (const auto& sv : svProd) {
+            NSV++;
+
+            GlobalVector flightDir(sv.vertex().x() - pv.x(), sv.vertex().y() - pv.y(), sv.vertex().z() - pv.z());
+            GlobalVector jetDir(pjet->px(), pjet->py(), pjet->pz());
+            if (reco::deltaR2(flightDir, jetDir) < 0.09) {
+                Measurement1D dl = vdist.distance(
+                                                  pv, VertexState(RecoVertex::convertPos(sv.position()), RecoVertex::convertError(sv.error())));
+                if (dl.significance() > maxFoundSignificance) {
+                    maxFoundSignificance = dl.significance();
+                    Jet_vtxPt = sv.pt();
+                    Jet_vtxMass = sv.p4().M();
+                    Jet_vtx3dL = dl.value();
+                    Jet_vtx3deL = dl.error();
+                    Jet_vtxNtrk = sv.numberOfSourceCandidatePtrs();
+                }
+            }
+        }
+        //variables needed for regression
+        Jet_pt = fjet.pt();
+        Jet_eta = fjet.eta() ;
+        Jet_leadTrackPt = leadTrackPt_; 
+        edm::Handle<double> rhoHandle;
+        evt.getByToken( rhoToken_, rhoHandle );
+        const double rhoFixedGrd = *( rhoHandle.product() );
+        rho = rhoFixedGrd;
+        Jet_mt = sqrt(fjet.energy()*fjet.energy()-fjet.pz()*fjet.pz());
+
+        //this max probably not needed, it's just heppy
+        Jet_leptonPtRel = std::max(float(0.),softLepPtRel);
+        Jet_leptonDeltaR = std::max(float(0.),softLepDr);
+        Jet_neHEF = fjet.neutralHadronEnergyFraction();
+        Jet_neEmEF = fjet.neutralEmEnergyFraction();
+        Jet_chHEF = fjet.chargedHadronEnergyFraction(); 
+        Jet_chEmEF = fjet.chargedEmEnergyFraction();
+
+        Jet_leptonPtRelInv = std::max(float(0.),softLepPtRelInv);
+        Jet_mass=fjet.mass();
+
+        float sumWeight=0;
+        float sumPt=0;
+        for(const auto & d : pjet->daughterPtrVector()){
+            sumWeight+=(d->pt())*(d->pt());
+            sumPt+=d->pt();
+        }
+        Jet_withPtd = (sumWeight > 0 ? sqrt(sumWeight)/sumPt : 0);
+
+    
+        
+
+        Jet_energyRing_dR0_em_Jet_e   = emEnergies[0]/fjet.energy();//remember to divide by jet energy
+        Jet_energyRing_dR1_em_Jet_e   = emEnergies[1]/fjet.energy();
+        Jet_energyRing_dR2_em_Jet_e   = emEnergies[2]/fjet.energy();
+        Jet_energyRing_dR3_em_Jet_e   = emEnergies[3]/fjet.energy();
+        Jet_energyRing_dR4_em_Jet_e   = emEnergies[4]/fjet.energy();
```

Comment:
Storing these variables in a ```std:;array``` would be shorter and less error-prone.

Reply:
True, again just kept the approach of flashgg

Reply:
This does not seem like a necessary change, so leaving it as is for now (I may come back on this at a later point)